### DISCUSSION TITLE
LFX Mentorship 2023 01-Mar-May Challenge - For #2226 and #2241

### DISCUSSION THREAD

**Discussion by hydai** [WasmEdge/WasmEdge#2227] [Show and tell]
### Introduction

For those who want to apply for LFX mentorship for #2226 and #2241, this is a pretest for the application.
The mission of this LFX mentorship is to implement a unified WasmEdge tool to meet the basic requirements; we designed this challenge to evaluate the candidates.

### Requirements

To complete this challenge, you'll need to create a standalone tool that uses [WasmEdge C SDK](https://wasmedge.org/book/en/sdk/c.html) to execute a given Wasm application.

The tool should support the following commands:
```bash
tool [version] [run] [wasm path] [arguments]
`version` is optional and prints the used wasmedge version; if this is given, you can ignore any other options.
`run` is optional.
`wasm path` is required; the relative and absolute path should be supported.
`arguments` is optional; some of the hidden wasm applications may not have arguments.
```

We will run the following command to check the tool:

```bash
# Assuming your tool is called `x`, and the given wasm application is called `add.wasm`, which will add two given numbers in the arguments
$ ./x version
0.11.2
$ ./x add.wasm 3 4
7
$ ./x run add.wasm 5 6
11
```

Once you are finished, please provide your link to the GitHub repository, which must contain a README file to show how you do and some screenshots of the examples' execution results.

### Deadline

Please refer to [the timeline of LFX mentorship 2023 01-Mar-May](https://github.com/cncf/mentoring/tree/main/lfx-mentorship/2023/01-Mar-May).
Since we have to process the internal review and decide, the final date for valid submissions will be February 20, 5:00 PM PDT.

### Reference

* WasmEdge C SDK: https://wasmedge.org/book/en/sdk/c.html

Good luck!

**Comment by VaibhavMalik4187** [comment]
Hi, I've tried to implement the required tool. The link to the repository is: https://github.com/VaibhavMalik4187/challenge
Please provide your feedback on my approach. Thanks!

**Reply by michael1017** [reply]
@hydai @VaibhavMalik4187 I don't think this is a coincidence.
https://github.com/VaibhavMalik4187/challenge/blob/e5fb731a456f722669bf1829890eb538a53885c5/main.c
https://github.com/michael1017/LFX_Mentorship_2023/blob/dc8aefb286b023f643486f82478f0e7877052ca1/src/include/baseutil.h
https://github.com/michael1017/LFX_Mentorship_2023/blob/dc8aefb286b023f643486f82478f0e7877052ca1/src/lib/handler.c

Base on the commit history, @VaibhavMalik4187 made his change in main.c after I released my work. Therefore, I can clearly say that @VaibhavMalik4187 completed his challenge based on my work, even though he made lots of changes afterward.

**Reply by VaibhavMalik4187** [reply]
@michael1017 First things first, I'm very grateful to you for helping me learn something new. You've worked very hard on this pretest. As for me, initially I submitted my project assuming that we have to create a tool that'll use the WasmEdge C SDK and it should be able to run add.WASM file. Later Mashijams submitted his work, I went through his approach and found a similar one. So I thought that I was headed in the right direction. Everyone until you submitted your pretest were using similiar approach as mine(hard-coding the function which was to be executed).
Your submission was great, looking at your solution I realized that what was actually missing in my approach. The add.wasm application was an example to demonstrate the use of the required tool to us. It did not have to be hard-coded for add.wasm only. Hence, I decided to correct my mistake and learn more about the Wasm C SDK. In the process I also learned to use CMake.
I also mentioned about this in my LFX cover letter(not in this much detail but yes a mention is there).
This was not only a pretest for me, I've learned very cool tech in this process. I also learned to use the Google Test suite in my project.
I took reference your code only to learn from it not to copy/paste the solution, that's all there is to it.

**Comment by sleipbyte** [comment]
Hello!

I completed this pretest.

To make this challenge more fun I added support for factorial.wasm, fibonacci.wasm along with add.wasm.

We can add even more wasm application support for this tool by just creating new derived classes in tool implementation (More info is in the Readme.md file).

Link to Repository: https://github.com/Mashijams/lfx_pretest

### Build and Tested On

#### OS: MacOSX
#### Arch: x86_64

### Brief Description

I named the tool **shell**, it interacts with different wasm applications and creates an instance of one as per arguments passed by the user.

For example: 
1. `./shell add.wasm 5 7` will create an instance of **Add** derived class
2. `./shell factorial.wasm 5` will create an instance of **factorial** derived class

Currently, shell supports add.wasm, fibonacci.wasm, factorial.wasm applications

All conditions in arguments for **version** and **run** are taken care of.

More implementation details can be found in the above repository.

It was a really fun task, do provide feedback for my work.

Thanks
Mashijams

**Comment by zyongye** [comment]
Hello, I complete this pretest and you can find them here https://github.com/zyongye/wasmedge-tool. Please provide any feedback if there is any. Thanks!!

**Comment by michael1017** [comment]
Hi @hydai , I have a question about this challenge.

Now I can use `WasmEdge_VMGetFunctionList` in WasmEdge C SDK to get the Wasm application’s function names. However, if the application has more than one function, how should we decide which function should be executed?

**Reply by hydai** [reply]
The execution of Wasm will need to specify the entry point function, so users will need to call the function by name [like this](https://github.com/WasmEdge/WasmEdge/blob/master/lib/driver/runtimeTool.cpp#L285)

**Comment by Tweoss** [comment]
Here's my submission: [tweoss/wasmedge_c_sdk_challenge](https://github.com/Tweoss/wasmedge_c_sdk_challenge). Notable features are executing any .wasm file, parameter handling for numeric types, and specifying which function is to be run. I'd appreciate feedback from whoever is willing to give theirs 🙁

**Comment by michael1017** [comment]
Hello, I have completed the pretest. The submission can be found in [michael1017/LFX_Mentorship_2023](https://github.com/michael1017/LFX_Mentorship_2023). Please feel free to provide any feedback if there is any.

**Comment by playerEric1** [comment]
Hello! I have completed testing the tool with add.wasm. Here is my submission [playerEric1/LFX-Pretest](https://github.com/playerEric1/LFX-Pretest). I will appreciate any feedback!

**Comment by dracoooooo** [comment]
Hi, I have completed the pretest. The repository link is https://github.com/dracoooooo/WasmTool. Please let me know if you have any feedback. Thanks.

**Comment by zriqo** [comment]
I have just finished pretest, kinda confusing at first. 
GitHub repo link https://github.com/lala-lala-lori/wasmtool

Hope I'll get the opportunity to learn more of wasm & wasmedge >.<

**Comment by Vincent1218** [comment]
Hello! I have completed the pretest and have uploaded my code to the following repository: https://github.com/Vincent1218/LFX_Wasmedge_tool
I have included a README.md file and comments in the code for easier reading and understanding.
Thank you for your time!

**Comment by Subhra264** [comment]
Greeting! Have finished the pretest. Check out the tool here - https://github.com/Subhra264/WasmRunner. Feel free to share any comments or suggestions regarding my submission, will be very happy to get feedback from anyone in the community. Thank you!
